boa_engine/src/optimizer/control_flow_graph/mod.rs

```diff
@@ -0,0 +1,549 @@
+
+    /// Finalize bytecode.
+    #[must_use]
+    pub fn finalize(self) -> Vec<u8> {
+        let index_from_basic_block = |bb: &RcBasicBlock| {
+            for (i, basic_block) in self.basic_blocks.iter().enumerate() {
+                if Rc::ptr_eq(basic_block, bb) {
+                    return i;
+                }
+            }
+
+            unreachable!("There should be a basic block")
+        };
+
+        let mut results = Vec::new();
+        let mut labels = Vec::new();
+        let mut blocks = Vec::with_capacity(self.basic_blocks.len());
+
+        for basic_block in &self.basic_blocks {
+            let basic_block = basic_block.borrow();
+
+            blocks.push(results.len() as u32);
+
+            for instruction in &basic_block.instructions {
+                instruction.to_bytecode(&mut results);
+            }
+
+            match &basic_block.terminator {
+                Terminator::None => {}
+                Terminator::JumpUnconditional { opcode, target } => {
+                    results.extend_from_slice(&[*opcode as u8]);
+                    let start = results.len();
+                    results.extend_from_slice(&[0, 0, 0, 0]);
+
+                    let target = index_from_basic_block(target);
+                    labels.push((start as u32, target));
+                }
+                Terminator::JumpConditional { opcode, no: _, yes } => {
+                    results.extend_from_slice(&[*opcode as u8]);
+                    let start = results.len();
+                    results.extend_from_slice(&[0, 0, 0, 0]);
+
+                    let target = index_from_basic_block(yes);
+                    labels.push((start as u32, target));
+                }
+                Terminator::TemplateLookup { yes, site, .. } => {
+                    results.extend_from_slice(&[Opcode::TemplateLookup as u8]);
+                    let start = results.len();
+                    results.extend_from_slice(&[0, 0, 0, 0]);
+                    results.extend_from_slice(&site.to_ne_bytes());
+
+                    let target = index_from_basic_block(yes);
+                    labels.push((start as u32, target));
+                }
+                Terminator::Return { .. } => {
+                    results.push(Opcode::Return as u8);
+                }
+            }
+        }
+
+        for (label, block_index) in labels {
+            let address = blocks[block_index];
+
+            let bytes = address.to_ne_bytes();
+            results[label as usize] = bytes[0];
+            results[label as usize + 1] = bytes[1];
+            results[label as usize + 2] = bytes[2];
+            results[label as usize + 3] = bytes[3];
+        }
+
+        results
+    }
+}
+
+impl Drop for ControlFlowGraph {
+    fn drop(&mut self) {
+        // NOTE: Untie BasicBlock nodes, so they can be deallocated.
+        for basic_block in &self.basic_blocks {
+            *basic_block.borrow_mut() = BasicBlock::default();
+        }
+    }
+}
```

Comment:
Could be good to see if we can use https://lib.rs/crates/slotmap, to avoid doing things like this.

Reply:
Will check it out! Thank you for the suggestion! :)